### PR TITLE
Convert all application icons to Material Design

### LINKED ISSUE #475: Improve application icons [$50]
I apologize for the somewhat light-minded request, but… no plan to switch to vector icons? On macOS, at least, they do not look so gorgeous… (the app icon is beautiful, though).

I know that resources are tight and there are other priorities. A good-looking UI, however, may attract a wider public to KeePassXC, making it eventually more popular.

### PR DESCRIPTION
**Pull request by wolframroesler**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)

... if you consider icon files to be "code". Only trivial changes to the C++ code were done.

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )

Convert all application icons (open database, create entry, database settings, etc. etc.) to "Material Design". Icons are taken from the Material Design repo (https://github.com/Templarian/MaterialDesign).

That means that all application icons are SVG now, rather than PNG as before. All fixed-size application icons have been replaced.

Includes a shell script to easily re-create the icon files from the Material Design repo (to be used e. g. when the Material Design repo has undergone changes that we want to incorporate into KeePassXC).

Database icons (the "default icons" you chose for entries in your password database) are unchanged.

Fixes #475

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )

Screenshot from 2019-12-27 15-26-04
Screenshot from 2019-12-27 15-26-28
Screenshot from 2019-12-27 15-26-43
Screenshot from 2019-12-27 15-26-52

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Tested manually. To facilitate testing, introduced a new environment variable that makes KeePassXC ignore the system icon theme and uses built-in icons only:

```
$ KEEPASSXC_IGNORE_ICON_THEME=1 keepassxc
```

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

**Comment by droidmonkey** [comment]
Does this also completely fix fuzzy icons on HiDPI? That is another bug we have open.

**Comment by droidmonkey** [comment]
You also need to update COPYING.md

**Comment by wolframroesler** [comment]
> Does this also completely fix fuzzy icons on HiDPI? That is another bug we have open.

@droidmonkey I hope so since we're using nothing but SVG now, however I can't try it out since I've got a regular monitor only.

**Comment by droidmonkey** [comment]
I'll be able to test that

**Comment by louib** [comment]
@wolframroesler really nice, thanks for that PR!

**Comment by droidmonkey** [comment]
Similar to the group icons, can we use the non-filled in icons for the entry menu? It looks a little off balance.

**Comment by droidmonkey** [comment]
The icons looks fuzzy on high DPI, perhaps we are missing a Qt setting somewhere.

image

**Comment by droidmonkey** [comment]
Simple fix, we need to set `AA_UseHighDpiPixmaps` everywhere else now. Please remove the $ifdef Q_OS_LINUX guards in main.cpp, line 53-55.

image

**Comment by wolframroesler** [comment]
By the way, there's nothing we can do to avoid that "issue" found by CodeFactor (nothing that's easy and not too convoluted I mean). Is there a way to say "this is fine, just ignore it"? Or do we just live with the message?

**Comment by wolframroesler** [comment]
> Similar to the group icons, can we use the non-filled in icons for the entry menu? It looks a little off balance.

Most icons come with an "outline" version, which I'm now using consistently. Looks much better, less heavy.

toolbar

entries

**Comment by droidmonkey** [comment]
I can ignore the code factor issue, no biggie

**Comment by varjolintu** [comment]
Tried this briefly today with macOS High Sierra in a non-HiDPI enviroment. The icons look great, but sometimes it's hard to distinguish them from each other. Especially the View or Edit Entry and Delete Entry icons. Open Database icon is also a bit strange. So multiple times I noticed I had to see what the icon tooltip says (I hardly use the icons so I don't remember the exact position and function of them).

**Comment by wolframroesler** [comment]
@varjolintu I'm not particularly happy with the icons you mentioned either, especially those from the "credit-card" family that I used for the entry menu, however it's not easy finding something supported for all the functions we need (create, view/edit, delete, clone). How about this ("comment" family)?

comment

In the icon maker script, this is achieved with the following:

```sh
        entry-clone)                        echo comment-multiple-outline       ;;
        entry-delete)                       echo comment-remove-outline         ;;
        entry-edit)                         echo comment-edit-outline           ;;
        entry-new)                          echo comment-plus-outline           ;;
```

There's also the "file-document-box" family, however "view/edit" doesn't have a pencil in it.

box

In the script that would be

```sh
        entry-clone)                        echo file-document-box-multiple-outline;;
        entry-delete)                       echo file-document-box-minus-outline;;
        entry-edit)                         echo file-document-box-outline      ;;
        entry-new)                          echo file-document-box-plus-outline ;;
```

I like the "comment" ones better despite the "speech bubble" triangle at the bottom. Opinions?

For the database icons, well ... there are the generic "folder" icons which come with "open" and "new" versions, but IMHO they look too generic and too similar to the entry and save icons:

database

Also, they don't come with a "merge" ("arrow into") version so we'd have to use a more generic icon for "merge databases" (cf. #122). I admit the "barrel" database icons may take a bit of getting used to, but I still like them better. Anyway, it would be:

```sh
        database-merge)                     echo merge                          ;;
        document-new)                       echo folder-star-outline            ;;
        document-open)                      echo folder-open-outline            ;;
```

TL;DR: I suggest keeping the database icons and switching entry icons like so:

main

**Comment by droidmonkey** [comment]
The folder open icon is very nice in your alternate toolbar. I think that with the barrel database icon for "new database" would be a good combo. Since "OPEN" is not typically associated with a barrel database icon, but the PLUS sign on the barrel makes sense for NEW. 

I like the comment icons.

**Comment by wolframroesler** [comment]
@droidmonkey How about using a generic "plus" icon for "new database"? That avoids the barrel altogether and goes nicely with generic "open" (folder) and generic "merge" (two-into-one arrow). Like so:

main

menu

**Comment by droidmonkey** [comment]
I am not a fan of that, it can be confused for "Add new entry" since its so generic. Although honestly I am not a fan of the new database button in the toolbar to begin with. It is not a function that is done often (if ever) and it is a dedicated button on the welcome screen for new users.

**Comment by phoerious** [comment]
I agree with the assessment. I would also prefer keeping the barrel icons and cleaning up the toolbar. New is not needed there, more are the settings IMHO.
We could even get rid of Save and only show that icon if auto save is off (or an error occurred).

**Comment by wolframroesler** [comment]
Will remove "new database" from the tool bar. In fact I never ever used it.

I do use the settings icon in the tool bar, however :) I have some other apps here that have a settings icon in plain sight all the time, which I find very convenient. Anyone willing to make the tool bar configurable?

Hiding the "Save" icon while Auto-Save is on and working sounds good but should go into a different issue.

**Comment by wolframroesler** [comment]
That was easy :) so the tool bar now looks like this:

main

So, what else shall we change? The three entry icons (create/edit/delete) should definitely stay, as should copy user name/password/URL and auto type (however I'd add a separator before auto type). I never use the "password generator" and "lock databases" icons (argh, and their order in the Tools menu is different). Also, since the Material Design icons are SVG, bigger icons (say, 32x32) are now possible; there seems to be some vertical space left due to the height of the search field, and I think it would look awesome with fewer but bigger icons.

OTOH, this PR is about re-designing icons, not re-designing the tool bar. We shouldn't pack too much into it, so let's add these ideas to #3585 instead.

**Comment by varjolintu** [comment]
> That was easy :) so the tool bar now looks like this:

It looks much better! :)

**Comment by droidmonkey** [comment]
We can do a general toolbar and menu overhaul in a new PR

**Comment by phoerious** [comment]
If we don't want to remove it, we could perhaps replace the settings icon with a less obtrusive hamburger icon and place it somewhere else. Can we do CSDs in Qt?

**Comment by wolframroesler** [comment]
> We can do a general toolbar and menu overhaul in a new PR

#3585, or yet another one?

**Comment by droidmonkey** [comment]
That is an issue, I mean a new pull request code submission. It would be linked to that issue, yes.

**Comment by wolframroesler** [comment]
> That is an issue, I mean a new pull request code submission. It would be linked to that issue, yes.

Ah right, sorry. I started a branch on my fork for #3585 already, will submit a PR when it has achieved some maturity.

@phoerious following up to your comment in #3585.

**Comment by wolframroesler** [comment]
OK let's discuss the other toolbar questions in #3585 but the new icons really look a lot better in 32x32 than the previous 22x22.

22x22
32x32

And 32x32 makes the toolbar just two or three pixels higher, a cheap price to pay for a big benefit IMHO. So, let's do it that way.

vertsize

**Comment by wolframroesler** [comment]
As requested in #3585, here are some alternative suggestions for the auto-type icon (each with the full toolbar for context).

* `auto-fix`, the "magic wand" that I'm currently using. Symbolizes the "automagical" log-on feature.
auto-fix

* `keyboard-outline` and `keyboard-settings-outline` are obvious choices, but I find the "stuff in a box" design too similar to the dice and lock icons.
keyboard-outline
keyboard-settings-outline

* `keyboard-variant` doesn't have a frame and is immediately recognized as a keyboard, also it's closest to the existing auto-type icon.
keyboard-variant

* `keyboard-tab` and `keyboard-return` are the tabulator and return key symbol, respectively, symbolizing the key codes that are sent by auto-type.
keyboard-tab
keyboard-return

* `auto-upload` The "A" could mean "auto" as well as a key getting pushed.
auto-upload

My preference is `keyboard-variant`. Your opinions?

**Comment by phoerious** [comment]
I generally prefer the keyboard variant as well, but the icons have very inconsistent heights, which gives a rather "wiggly" look.

**Comment by wolframroesler** [comment]
Finding alternatives for "copy user name/password/URL to clipboard" isn't so easy. I agree that the current ones merely symbolize "user name/password/URL" without the "copy to clipboard" part; the tricky part is to find alternatives that symbolize "copy" or "to clipboard", and of course all three in the same way. There is `account-arrow-right-outline`

account-arrow-right-outline

which could be understood as "put user somewhere" aka "copy user name to clipboard", but we don't have "arrow pointing away from" versions of the key and earth icons. Neither did I find anything that could symbolize password or URL other than key and earth. There are many variations of clipboards but none of them contains anything that suggests user names, passwords, or URLs.

The only idea I have is to request new icons from the Material Design maintainers, use them as soon as they become available, and live with the current choice until then.

**Comment by phoerious** [comment]
We could perhaps reduce the three buttons to a single copy button (I assume that's the most common use case). Then we could add a little dropdown menu to the button that contains the three explicit options, whereas clicking the button itself would only copy the password. This would also be in line with the Ctrl+C shortcut, which also only copies the password.

**Comment by droidmonkey** [comment]
I personally like `keyboard-settings-outline` for Auto-Type. The dots below make me think of typing a password.

**Comment by wolframroesler** [comment]
> We could perhaps reduce the three buttons to a single copy button (I assume that's the most common use case). Then we could add a little dropdown menu to the button that contains the three explicit options, whereas clicking the button itself would only copy the password. This would also be in line with the Ctrl+C shortcut, which also only copies the password.

Don't like that at all. Simple things should be simple, and every click hurts. I don't think that doubling the number of clicks required to copy something to the clipboard is desirable (even if doubling from one to two clicks). Also, a dropdown menu with the three copy functions exists already (the "Entries" menu); with your suggestion we'd simply duplicate part of that menu into the toolbar. Also, there's Ctrl-B and Ctrl-U in addition to  Ctrl-C, and IMO they're in line with the way the toolbar currently works (single key press vs. single click).

**Comment by phoerious** [comment]
My intention is just to simplify the toolbar and remove buttons that are rarely used.

**Comment by droidmonkey** [comment]
@wolframroesler I rebased onto develop, would like to squash all these intermediary commits as well, ok?

**Comment by wolframroesler** [comment]
@droidmonkey Squashing is fine by me, but please scan the comments first to check if there's anything worth keeping. I think I put the source URL of the icons somewhere (don't know if it's needed in the comments, however, because it's in the copying file also).

**Comment by phoerious** [comment]
While we are at it... You should create an index.theme file for the new icon theme and compile the whole icon theme into the binary using the Qt resource framework. This way we can get rid of all the logic for guessing paths in core/FilePath.cpp and we also fix #2582.

**Comment by wolframroesler** [comment]
Requested "key with outgoing arrow" and "earth with outgoing arrow" (for "copy password to clipboard" and "copy URL to clipboard", respectively) in the Material Design repo (https://github.com/Templarian/MaterialDesign/issues/4729).

**Comment by wolframroesler** [comment]
> While we are at it... You should create an index.theme file for the new icon theme and compile the whole icon theme into the binary using the Qt resource framework. This way we can get rid of all the logic for guessing paths in core/FilePath.cpp and we also fix #2582.

Sounds good. For reference: https://openapplibrary.org/dev-tutorials/qt-icon-themes

However, shouldn't that rather get its own issue and PR? The point is that it will require quite a few changes to the C++ code, which the current PR leaves completely unchanged. Also, I suppose the icon theme should include the database icons as well, and this issue (#475) and PR are about the application icons only. So, I suggest to finish the application icons first and then make the icon theme, either before or after redesigning the database icons (#4071).

**Comment by phoerious** [comment]
Can we get a preliminary version of this PR ready so I can base my theming branch on it already?
Further adjustments and the Qt resource migration can then be followed up with in another PR.

**Comment by wolframroesler** [comment]
Incredible, the copy password/URL icons are finished! https://github.com/Templarian/MaterialDesign/issues/4729#issuecomment-573273083
Will add them to the PR ASAP.

**Comment by wolframroesler** [comment]
And here's the KeePassXC logo on Simple Icons: https://github.com/simple-icons/simple-icons/commit/9b41018d3223ccc5c29fecb792cad187f451b3b1?short_path=b21461e#diff-b21461edeb4f19851047f504c729d397

For reference: https://github.com/simple-icons/simple-icons/issues/2268

@phoerious Will integrate these new icons ASAP, then I'm done for now, and you can go ahead. What kind of preliminary version of the PR do you need? Can I be of help creating one?

**Comment by phoerious** [comment]
One that we can merge already. ;-)

**Comment by wolframroesler** [comment]
Add the new "copy to clipboard" icons. Here's how the toolbar looks now:

toolbar

So, @phoerious, is there anything left to do before it can be merged?

**Comment by phoerious** [comment]
Thanks. I squashed the commits.

**Comment by droidmonkey** [comment]
Nice squash job phoerious

**Comment by phoerious** [comment]
Thank you. Here's a screenshot for you:

image

I noticed that the "Open" icon is not very suited for low-DPI screens, particularly the version in the "Database" menu, which is even smaller. Can we find something better and more in line with the pixel grid?

image

**Comment by droidmonkey** [comment]
Perhaps it needs to be tweaked from the source?

**Comment by phoerious** [comment]
The folder tab section looks blurry and needs to be adjusted to fit a regular raster.

**Comment by wolframroesler** [comment]
Maybe it's related not to the icon (which is SVG) but to scaling? How about other icons, are they blurry as well?

Alignment is OK for me:

Screenshot from 2020-01-11 19 10 09

**Comment by phoerious** [comment]
Quite possible. That's why I asked for PNGs to be provided with the correct resolution (for icons of which we do not have SVGs) instead of relying on Qt to scale them properly. Sometimes there is even a dedicated 16x16 version of an icon and a different one for 32x32 or 48x48 etc. I had hoped the SVG rendering works better, but maybe it doesn't and we also need to compile PNGs from them.

**Comment by wolframroesler** [comment]
Hm, all application icons are SVG now, because that's all we get from the Material Design repo. Maybe it would help to pre-create PNGs in various resolutions with Image Magick or a similar tool and ship these instead of the SVGs. It would be quite easy to do that in a shell script.

How does that go with your Qt theme, @phoerious?

**Comment by phoerious** [comment]
Wouldn't be any different, you only have to add the PNGs to the qrc file.

**Comment by wolframroesler** [comment]
Working on a prototype, will submit a first version soon.

#475 is closed, is there a more suitable issue to reference in the new commits?

**Comment by wolframroesler** [comment]
I  wrote a shell script that converts the SVGs to PNGs in various resolutions. I'm not quite satisfied, however, because the result looks more blurry than the SVG version, at least on my system:

SVG version
with-svg

PNG version
with-png

The script contains three implementations of SVG-to-PNG conversions, none of which seem to work perfectly. However, judging if the PNGs are crisp or blurry pushes my eyesight (or my monitor, or both) to the limits. @droidmonkey maybe you can pull it and have a look yourself. I really hope we find a solution that doesn't require the icons to be hand-tailored for low resolutions.

And probably this should go into a new issue.

**Comment by droidmonkey** [comment]
I am not a fan of prebaking PNG. If an SVG renders incorrectly, fix the SVG.

**Comment by wolframroesler** [comment]
@phoerious 

> That being said, you can redesign SVG icons to improve low-res pixel rasterisation by aligning them with a grid, avoiding different stroke widths (like the one used for the folder tab), and lines with angles other than 90 or 45 degrees.

I'm afraid this is beyond my area of expertise. Maybe you can report it to the Material Design guys (https://github.com/Templarian/MaterialDesign)? They have been extremely responsive. The name of the icon in question is `folder-open-outline`.

**Comment by phoerious** [comment]
@wolframroesler How do we handle dark themes and selected states of icons? In my PR #4110 I have an ugly intermediary solution where I convert the SVGs to 128x128 pixmaps and then recolour them a) for inverted icons in selection highlights and b) for dark themes (which I try to detect based on the standard palette).

I think there might be a more elegant way to achieve this which does not include creating fixed-size pixmaps nor needing two different icon sets.

Also an interesting observation: Providing only 128x128 icons and then letting Qt do the rest works better than creating multiple sizes beforehand. Perhaps that's because some icons are not exactly 16x16 or so and scaling them to the correct size from 128 produces better results than scaling from the nearest already-low resolution, who knows.

**Comment by wolframroesler** [comment]
@phoerious replying in #4110.

**Comment by OLLI-S** [comment]
@wolframroesler Regarding these icons in https://github.com/keepassxreboot/keepassxc/pull/4066#issuecomment-573357634:

image

Are these icons final?
The icons "Add", "Edit" and "Delete" have a "Speech Bubble" icon, but you are programming a password manager, not a chat client (I did not find any chat features yet **;-)** ).

**Comment by Saroumane** [comment]
Hello, my distro just upgraded to KeepassXC 2.6.x and I come here to see what happened.
I don't understand the rationale behind going from colorful application icons to monochrome icons.
As previous users have highlighted in this thread, it causes an extra effort to avoid confusion between monochrome icons.
That's why I think it's a step backwards in terms of usability.

Could you please provide a setting to restore colorful unambiguous applications icons ?

**Comment by phoerious** [comment]
Not planned, so no.
